controller/object_store_controller.go

```diff
@@ -0,0 +1,870 @@
+package controller
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/kubernetes/pkg/controller"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+var (
+	ErrObjectStorePVCNotReady        = errors.New("PVC not ready")
+	ErrObjectStoreVolumeNotReady     = errors.New("Volume not ready")
+	ErrObjectStorePVNotReady         = errors.New("PV not ready")
+	ErrObjectStoreDeploymentNotReady = errors.New("Deployment not ready")
+	ErrObjectStoreServiceNotReady    = errors.New("Service not ready")
+	ErrObjectStoreIngressNotReady    = errors.New("Ingress not ready")
+)
+
+type ObjectStoreController struct {
+	*baseController
+
+	namespace string
+	ds        *datastore.DataStore
+	s3gwImage string
+	uiImage   string
+
+	cacheSyncs []cache.InformerSynced
+}
+
+func NewObjectStoreController(
+	logger logrus.FieldLogger,
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubeClient clientset.Interface,
+	controllerID string,
+	namespace string,
+	objectStoreImage string,
+	objectStoreUIImage string,
+) *ObjectStoreController {
+	osc := &ObjectStoreController{
+		baseController: newBaseController("object-store", logger),
+		namespace:      namespace,
+		ds:             ds,
+		s3gwImage:      objectStoreImage,
+		uiImage:        objectStoreUIImage,
+	}
+
+	ds.ObjectStoreInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    osc.enqueueObjectStore,
+			UpdateFunc: func(old, cur interface{}) { osc.enqueueObjectStore(cur) },
+			DeleteFunc: osc.enqueueObjectStore,
+		},
+	)
```

Comment:
Suggesting to disable the resyncPeriod and add more informer (Volume, Deployment, Service,...). resyncPeriod increase the CPU load because by default we resync every ObjectStore objects for every 30s even if there is nothing change in the system. See example at snapshot-controller [link](https://github.com/longhorn/longhorn-manager/blob/5e0868b84aca3cc601ad00cc0e37f8bf01d3efd6/controller/snapshot_controller.go#L78-L92)

Reply:
This is a good idea. I tried it out with just monitoring the ObjectStores as it is here and Deployments like in the snapshot controller and it works good already. I do wonder if we should retain a resync period of like 30min or so, just to make sure the controller checks in on the state of the gateway every now and then.

Reply:
Sounds good to me to keep 1h resync resync for the `ObjectStoreInformer` just in case some event was missing. We don't need the resyncPeriod for `DeploymentInformer` and `VolumeInformer`